gremlin-go/driver/traversal_test.go

```diff
@@ -62,4 +65,111 @@ func TestTraversal(t *testing.T) {
 		}, bytecode.stepInstructions[1].arguments[0])
 		assert.Equal(t, "binding[b=created]", bytecode.stepInstructions[1].arguments[0].(*Binding).String())
 	})
+
+	t.Run("Test Transaction commit", func(t *testing.T) {
+		skipTestsIfNotEnabled(t, integrationTestSuiteName, testTransactionEnable)
+		// Start a transaction traversal.
+		remote := newConnection(t)
+		g := Traversal_().WithRemote(remote)
+		startCount := getCount(t, g)
+		tx := g.Tx()
+
+		// Except transaction to not be open until begin is called.
+		assert.False(t, tx.IsOpen())
+		gtx, _ := tx.Begin()
+		assert.True(t, tx.IsOpen())
+
+		addV(t, gtx, "lyndon")
+		addV(t, gtx, "valentyn")
+		assert.Equal(t, startCount, getCount(t, g))
+		assert.Equal(t, startCount+2, getCount(t, gtx))
+
+		// Commit the transaction, this should close it.
+		// Our vertex count outside the transaction should be 2 + the start count.
+		_, err := tx.Commit()
+		assert.Nil(t, err)
+
+		assert.False(t, tx.IsOpen())
+		assert.Equal(t, startCount+2, getCount(t, g))
+
+		dropGraphCheckCount(t, g)
+		verifyGtxClosed(t, gtx)
+	})
+
+	t.Run("Test Transaction rollback", func(t *testing.T) {
+		skipTestsIfNotEnabled(t, integrationTestSuiteName, testTransactionEnable)
+		// Start a transaction traversal.
+		remote := newConnection(t)
+		g := Traversal_().WithRemote(remote)
+		startCount := getCount(t, g)
+		tx := g.Tx()
+
+		// Except transaction to not be open until begin is called.
+		assert.False(t, tx.IsOpen())
+		gtx, _ := tx.Begin()
+		assert.True(t, tx.IsOpen())
+
+		addV(t, gtx, "lyndon")
+		addV(t, gtx, "valentyn")
+		assert.Equal(t, startCount, getCount(t, g))
+		assert.Equal(t, startCount+2, getCount(t, gtx))
+
+		// Rollback the transaction, this should close it.
+		// Our vertex count outside the transaction should be the start count.
+		_, err := tx.Rollback()
+		assert.Nil(t, err)
+
+		assert.False(t, tx.IsOpen())
+		assert.Equal(t, startCount, getCount(t, g))
+
+		dropGraphCheckCount(t, g)
+		verifyGtxClosed(t, gtx)
+	})
+}
+
```

Comment:
Can we add some additional testing here? Python has a number of additional tests, I think we should add them and enhance our coverage since this code can be tricky.

```
    @pytest.mark.skipif(transactions_disabled(), reason="Transactions are not enabled.")
    def test_transaction_no_begin(self, remote_transaction_connection):
        # Start a transaction traversal.
        g = traversal().withRemote(remote_transaction_connection)
        tx = g.tx()

        # Except transaction to not be open until begin is called.
        assert not tx.isOpen()

        try:
            # Attempt to commit even though no tx is started.
            tx().commit()
            assert False
        except Exception as e:
            assert not tx.isOpen()

        try:
            # Attempt to rollback even though no tx is started.
            tx().rollback()
            assert False
        except Exception as e:
            assert not tx.isOpen()

        try:
            # Attempt to invoke tx().tx() which is illegal.
            tx().tx()
            assert False
        except Exception as e:
            assert not tx.isOpen()

        # Create transaction and verify it is open.
        gtx = tx.begin()
        assert tx.isOpen()
        try:
            # Attempt to begin gtx which is illegal.
            gtx().begin()
            assert False
        except Exception as e:
            assert tx.isOpen()

        # Commit this unused transaction and verify it is no longer open.
        tx.commit()
        assert not tx.isOpen()

        # Create another transaction and verify it is open.
        gtx = tx.begin()
        assert tx.isOpen()

        # Rollback this unused transaction and verify it is no longer open.
        tx.rollback()
        assert not tx.isOpen()

    @pytest.mark.skipif(transactions_disabled(), reason="Transactions are not enabled.")
    def test_multi_commit_transaction(self, remote_transaction_connection):
        # Start a transaction traversal.
        g = traversal().withRemote(remote_transaction_connection)
        start_count = g.V().count().next()

        # Create two transactions.
        tx1, tx2 = g.tx(), g.tx()

        # Generate two GraphTraversalSource's for each transaction with begin.
        gtx1, gtx2 = tx1.begin(), tx2.begin()
        verify_tx_state([tx1, tx2], True)

        # Add node to gtx1, which should be visible to gtx1, not gtx2.
        add_node_validate_transaction_state(g, gtx1, start_count, start_count, [tx1, tx2])

        # Add node to gtx2, which should be visible to gtx2, not gtx2
        add_node_validate_transaction_state(g, gtx2, start_count, start_count, [tx1, tx2])

        # Add node to gtx1, which should be visible to gtx1, not gtx2. Note previous node also added.
        add_node_validate_transaction_state(g, gtx1, start_count, start_count + 1, [tx1, tx2])

        tx1.commit()
        verify_tx_state([tx1], False)
        verify_tx_state([tx2], True)
        assert g.V().count().next() == start_count + 2

        tx2.commit()
        verify_tx_state([tx1, tx2], False)
        assert g.V().count().next() == start_count + 3

    @pytest.mark.skipif(transactions_disabled(), reason="Transactions are not enabled.")
    def test_multi_rollback_transaction(self, remote_transaction_connection):
        # Start a transaction traversal.
        g = traversal().withRemote(remote_transaction_connection)
        start_count = g.V().count().next()

        # Create two transactions.
        tx1, tx2 = g.tx(), g.tx()

        # Generate two GraphTraversalSource's for each transaction with begin.
        gtx1, gtx2 = tx1.begin(), tx2.begin()
        verify_tx_state([tx1, tx2], True)

        # Add node to gtx1, which should be visible to gtx1, not gtx2.
        add_node_validate_transaction_state(g, gtx1, start_count, start_count, [tx1, tx2])

        # Add node to gtx2, which should be visible to gtx2, not gtx2
        add_node_validate_transaction_state(g, gtx2, start_count, start_count, [tx1, tx2])

        # Add node to gtx1, which should be visible to gtx1, not gtx2. Note previous node also added.
        add_node_validate_transaction_state(g, gtx1, start_count, start_count + 1, [tx1, tx2])

        tx1.rollback()
        verify_tx_state([tx1], False)
        verify_tx_state([tx2], True)
        assert g.V().count().next() == start_count

        tx2.rollback()
        verify_tx_state([tx1, tx2], False)
        assert g.V().count().next() == start_count

    @pytest.mark.skipif(transactions_disabled(), reason="Transactions are not enabled.")
    def test_multi_commit_and_rollback(self, remote_transaction_connection):
        # Start a transaction traversal.
        g = traversal().withRemote(remote_transaction_connection)
        start_count = g.V().count().next()

        # Create two transactions.
        tx1, tx2 = g.tx(), g.tx()

        # Generate two GraphTraversalSource's for each transaction with begin.
        gtx1, gtx2 = tx1.begin(), tx2.begin()
        verify_tx_state([tx1, tx2], True)

        # Add node to gtx1, which should be visible to gtx1, not gtx2.
        add_node_validate_transaction_state(g, gtx1, start_count, start_count, [tx1, tx2])

        # Add node to gtx2, which should be visible to gtx2, not gtx2
        add_node_validate_transaction_state(g, gtx2, start_count, start_count, [tx1, tx2])

        # Add node to gtx1, which should be visible to gtx1, not gtx2. Note previous node also added.
        add_node_validate_transaction_state(g, gtx1, start_count, start_count + 1, [tx1, tx2])

        tx1.commit()
        verify_tx_state([tx1], False)
        verify_tx_state([tx2], True)
        assert g.V().count().next() == start_count + 2

        tx2.rollback()
        verify_tx_state([tx1, tx2], False)
        assert g.V().count().next() == start_count + 2

    @pytest.mark.skipif(transactions_disabled(), reason="Transactions are not enabled.")
    def test_transaction_close_tx(self):
        remote_conn = create_connection_to_gtx()
        g = traversal().withRemote(remote_conn)

        drop_graph_check_count(g)

        tx1 = g.tx()
        tx2 = g.tx()

        # open up two sessions and create stuff
        gtx1 = tx1.begin()
        gtx2 = tx2.begin()

        add_node_validate_transaction_state(g, gtx1, 0, 0, [tx1, tx2])
        add_node_validate_transaction_state(g, gtx2, 0, 0, [tx1, tx2])
        add_node_validate_transaction_state(g, gtx2, 0, 1, [tx1, tx2])
        add_node_validate_transaction_state(g, gtx2, 0, 2, [tx1, tx2])

        # someone gets lazy and doesn't commit/rollback and just calls close() - the graph
        # will decide how to treat the transaction, but for neo4j/gremlin server in this
        # test configuration it should rollback
        tx1.close()
        tx2.close()

        assert not tx1.isOpen()
        assert not tx2.isOpen()
        verify_gtx_closed(gtx1)
        verify_gtx_closed(gtx2)

        remote_conn = create_connection_to_gtx()
        g = traversal().withRemote(remote_conn)
        assert g.V().count().next() == 0

        drop_graph_check_count(g)

    @pytest.mark.skipif(transactions_disabled(), reason="Transactions are not enabled.")
    def test_transaction_close_tx_from_parent(self):
        remote_conn = create_connection_to_gtx()
        g = traversal().withRemote(remote_conn)

        drop_graph_check_count(g)

        tx1 = g.tx()
        tx2 = g.tx()

        # open up two sessions and create stuff
        gtx1 = tx1.begin()
        gtx2 = tx2.begin()

        add_node_validate_transaction_state(g, gtx1, 0, 0, [tx1, tx2])
        add_node_validate_transaction_state(g, gtx2, 0, 0, [tx1, tx2])
        add_node_validate_transaction_state(g, gtx2, 0, 1, [tx1, tx2])
        add_node_validate_transaction_state(g, gtx2, 0, 2, [tx1, tx2])

        # someone gets lazy and doesn't commit/rollback and just calls close() but on the parent
        # DriverRemoteConnection for all the session that were created via tx() - the graph
        # will decide how to treat the transaction, but for neo4j/gremlin server in this
        # test configuration it should rollback
        remote_conn.close()

        assert not tx1.isOpen()
        assert not tx2.isOpen()
        verify_gtx_closed(gtx1)
        verify_gtx_closed(gtx2)

        remote_conn = create_connection_to_gtx()
        g = traversal().withRemote(remote_conn)
        assert g.V().count().next() == 0

        drop_graph_check_count(g)
```

Reply:
added in https://github.com/lyndonbauto/tinkerpop/pull/68/commits/e7ea1f2b7a6b8d3a9a150b2482a8e57a749bad2b
There is 1 issue left with closing the session

Reply:
tests passing now